### PR TITLE
feat: MoE QLoRA training, QAT, quantized AdamW, Critical token SFT, and expert pruning

### PR DESCRIPTION
**Pull request by DFveloper**
## Overview

This PR consolidates the MoE training, quantization-aware training, optimizer, training-pipeline, Critical SFT, and MoE pruning work into a single implementation.
Also, this PR is extension of https://github.com/ggml-org/llama.cpp/pull/22705 .

Main changes:

* QLoRA fine-tuning for Mixture-of-Experts models using the `MUL_MAT_ID` backward pass.
* Standard SFT and reward-weighted SFT / GRPO-style training.
* Quantized AdamW optimizers.
* Quantization-aware training with STE backward propagation.
* Critical SFT for selectively increasing loss weights on important tokens and spans.
* Cosine LR scheduling, configurable warmup, verbose loss reporting, and checkpoint resumption.
* Multithreaded JSONL parsing and training data-pipeline improvements.
* Experimental ROCm support.
* MoE calibration, expert-importance analysis, soft pruning, and experimental hard pruning.

The QLoRA implementation supports MoE architectures including Mixtral, Qwen-MoE, DeepSeek-MoE, and the supported Gemma 4 MoE configuration.

The pruning implementation currently targets Gemma 4 26B A4B, with hard pruning limited to the repository's expected Q4_0 routed-expert tensor layout.

### MoE QLoRA

Added:

```text
examples/qlora_training/finetune_qlora.cpp
```

The training implementation includes:

* `MUL_MAT_ID` backward support.
* `OUT_PROD_ID`.
* MoE reverse-propagation fixes.
* Standard supervised fine-tuning.
* Reward-weighted SFT / GRPO-style training.
* Gradient checkpointing to reduce activation VRAM usage.
* AdamW `gclip`.
* Gemma 4 MoE support.
* `check_lora_norms.py` and sample datasets for diagnostics and testing.

The QLoRA work includes functionality previously developed in the related `feat/backward-mul-mat-id` work and associated PRs.

### Quantized optimizers

Added experimental AdamW optimizer variants using:

* F16
* Q8_0
* Q6_K

Lower-precision configurations automatically use Q8_0 or F32 for the AdamW second-moment accumulator (`v`) when the selected representation does not provide sufficient precision.

This avoids underflow that can otherwise result in unstable or `NaN` loss.

Q8_0 scaling was also changed to use the full signed int8 range, including `-128`, instead of limiting the range to `±127`.

IQ4_NL optimizer support was tested experimentally but is not currently enabled.

### Quantization-aware training

Added QAT using straight-through estimator (STE) backward propagation.

Supported QAT formats currently include:

* `Q8_0`
* `Q6_K`
* `Q4_0`
* `MXFP4`
* `Q4_K`
* `Q3_K`

Implementations are included for CPU, CUDA, Vulkan, and Metal where applicable.

A Q4_K QAT issue involving flipped `scale` / `min` bits was fixed. In the development test case, normalized MSE decreased from approximately `0.5` to `3e-5`–`4e-5`.

### Critical SFT

Added Critical SFT for selectively assigning larger loss weights to important parts of the target response instead of weighting the entire response uniformly.

Supported modes:

```text
none
spans
confidence
hybrid
```

The implementation supports configurable:

* critical-token/span weight,
* confidence threshold,
* constant or linear weighting shape,
* warmup,
* maximum weighted fraction,
* periodic weighting statistics.

The `spans` mode operates on explicitly marked critical regions.

The `confidence` mode derives critical regions from the configured importance/confidence signal.

The `hybrid` mode combines both mechanisms.

This is intended for training data where reasoning steps, corrections, decisive answer tokens, or other high-information regions should have greater influence on the loss than surrounding routine text.

Initial development tests showed substantially higher token-level target accuracy on the tested distillation datasets, but these results are workload-specific and are not intended as general benchmark claims.

### LR scheduler and warmup

Added:

```text
--lr-scheduler constant|cosine
--warmup-steps N
--warmup-init-ratio R
-lr-min LR
```

`--warmup-steps` uses logical training steps.

For SFT, the scheduler advances once per dataset window.

For GRPO-style training, it advances once per training iteration.

Ubatch gradient accumulation does not double-count scheduler steps.

Scheduler state is saved as:

```text
training.scheduler.step
```

and restored during checkpoint resume.

`--warmup-init-ratio` defaults to `0.1` and accepts:

```text
0 < ratio <= 1
```

Invalid values, including zero, negative values, NaN, and values above `1`, are rejected.

Warmup uses:

```cpp
start_lr = peak_lr * warmup_init_ratio;
progress = float(step) / warmup_steps;
lr = start_lr + (peak_lr - start_lr) * progress;
```

Example:

```text
peak_lr=1e-4, warmup_steps=2, ratio=0.2

step 0: 2e-5
step 1: 6e-5
step 2: 1e-4
step 3: 5.5e-5  // cosine decay
```

Example usage:

```bash
--lr-scheduler cosine \
--warmup-steps 10 \
--warmup-init-ratio 0.1 \
-lr 1e-4 \
-lr-min 1e-6
```

### Verbose loss

Added:

```text
--verbose-loss
```

It is disabled by default and works independently of `--save-every`.

Example output:

```text
epoch=1 window=2 window_loss=10.407519 ema16=10.398861 ema64=10.398010 epoch_mean=10.402613 lr=6e-05
```

The output contains:

* `epoch`
* `window`
* `window_loss`
* `ema16`
* `ema64`
* `epoch_mean`
* current LR

`window_loss` is the mean ubatch loss for the current logical window.

EMA state is maintained across epochs, while `epoch_mean` resets each epoch.

This is intended primarily for training diagnostics and hyperparameter tuning.

### Checkpoint resumption

Added:

```bash
--resume /path/to/ckpt/name.epochE.ckptW.gguf
```

Relevant training state, including scheduler progress, is restored when resuming.

### Data pipeline

Added multithreaded JSONL parsing and CPU-side preprocessing through `-dthr`.

Conversation datasets can be processed using the model's Jinja template and tokenizer.

Supported token-mask modes include:

```text
all
assistant
reasoning
content
```

with token-start boundary handling for masked regions.

Gradient offload was also prototyped but remains disabled because the current implementation is not considered reliable enough.

### ROCm

Experimental ROCm support is included, primarily by reusing compatible CUDA implementation paths.

This support is currently experimental and may be changed, separated, or dropped depending on upstream backend requirements.

### MoE pruning

Added calibration-based MoE expert analysis and pruning tooling through:

```text
llama-prune
```

Commands currently include:

```text
llama-prune analyze
llama-prune inspect
llama-prune hard
```

Calibration collects per-layer and per-expert information including:

* routing counts and frequencies,
* router probability statistics,
* output norms,
* router-output importance,
* routing entropy,
* routing imbalance,
* invalid-routing statistics.

Multiple pruning ratios can be evaluated from the same calibration run.

Generated pruning plans are deterministic and nested, subject to per-layer and Top-K safety limits.

The analyzer can report:

* baseline perplexity,
* per-ratio perplexity,
* timing,
* throughput,
* routing entropy,
* routing imbalance,
* invalid-routing statistics.

Example:

```bash
build/bin/llama-prune analyze \
  --model model.gguf \
  --dataset calibration.jsonl \
  --ratios 0.05,0.10,0.20 \
  --ppl-mask assistant \
  --output-dir results
```

### Soft pruning

Analysis can generate versioned, hash-bound pruning profiles plus JSON, CSV, and text summaries.

Example profile:

```json
{
  "format": "llama-moe-prune",
  "version": 1,
  "mode": "soft",
  "model": {
    "architecture": "gemma4",
    "model_hash": "sha256:...",
    "expert_tensor_hash": "sha256:...",
    "expert_count": 128,
    "experts_used": 8
  },
  "calibration": {
    "dataset_hash": "sha256:...",
    "ppl_mask": "assistant",
    "metric": "router-output",
    "evaluated_tokens": 100000
  },
  "pruning": {
    "requested_ratio": 0.2,
    "actual_ratio": 0.1953125,
    "layers": {
      "0": {
        "disabled_experts": [3, 7, 11]
      }
    }
  }
}
```

Profiles can be loaded by `llama-server`:

```bash
build/bin/llama-server \
  --model model.gguf \
  --moe-prune-profile results/profile-020.json
```

Disabled experts are masked before grouped routing and Top-K selection.

### Hard pruning

Experimental hard pruning is implemented for Gemma 4 26B A4B with the repository's expected Q4_0 routed-expert layout.

The implementation includes:

* raw block-aligned expert copying,
* router compaction,
* metadata rewriting,
* expert-index remapping,
* shared-MLP preservation,
* explicit rejection of unsupported or unsafe layouts.

Output uses temporary staging.

The requested final path is replaced only after the generated model passes normal-loader reload and inference smoke validation.

Example:

```bash
build/bin/llama-prune hard \
  --model model.gguf \
  --profile results/profile-020.json \
  --output model-pruned.gguf \
  --dataset calibration.jsonl \
  --validate
```

A dataset can optionally be supplied for soft-versus-hard perplexity comparison.

A synthetic GGUF fixture covers:

* expert tensor slicing,
* router slicing,
* metadata rewriting,
* expert remapping,
* GGUF reopening.

Other architectures and unsupported Gemma 4 layouts fail explicitly rather than attempting unsafe conversion.

A production-size Gemma 4 26B A4B Q4_0 QAT GGUF is still required for final end-to-end validation of the hard-pruning path, including full calibration, hard conversion, normal-loader reload, smoke inference, and soft-versus-hard perplexity comparison.

For this reason, hard pruning should currently be considered experimental.

### Operator count

This training implementation adds an operator alongside recent upstream operator additions.

For the branch state this work was developed against:

```text
GGML_OP_COUNT = 102
```

is expected, and the corresponding `static_assert` was updated.

This may require adjustment during rebasing if upstream adds or removes operators.

## Requirements

* I have read and agree with the [[contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
* AI usage disclosure: YES. AI assistance was used for code search, translation, diff organization, routine mechanical edits, and applying already-decided repetitive changes consistently across backends. The implementation, technical decisions, testing, and final review were performed or independently verified by the contributors.

### CODE REVIEW

**Comment by 0cc4m** [comment]
Way too many changes
